actors/miner/src/expiration_queue.rs

```diff
@@ -363,64 +364,50 @@ impl<'db, BS: Blockstore> ExpirationQueue<'db, BS> {
-                .context("too many sectors to reschedule")?
-                .map(|i| i as SectorNumber)
-                .collect();
-
-            // This loop could alternatively be done by constructing bitfields and intersecting them, but it's not
-            // clear that would be much faster (O(max(N, M)) vs O(N+M)).
-            // If faults are correlated, the first queue entry likely has them all anyway.
-            // The length of sectors has a maximum of one partition size.
-            for sector in sectors.iter() {
-                let sector_number = sector.sector_number;
-                let power = power_for_sector(sector_size, sector);
-                let mut found = false;
-
-                if on_time_sectors.contains(&sector_number) {
-                    found = true;
-                    // If the sector expires on-time at this epoch, leave it here but change faulty power to active.
-                    // The pledge is already part of the on-time pledge at this entry.
-                    expiration_set.faulty_power -= &power;
-                    expiration_set.active_power += &power;
-                } else if early_sectors.contains(&sector_number) {
-                    found = true;
-                    // If the sector expires early at this epoch, remove it for re-scheduling.
-                    // It's not part of the on-time pledge number here.
-                    expiration_set.early_sectors.unset(sector_number);
-                    expiration_set.faulty_power -= &power;
-                    sectors_rescheduled.push(sector);
-                }
-
-                if found {
-                    recovered_power += &power;
-                    remaining.remove(&sector.sector_number);
-                }
+        for sector in sectors.iter() {
+            let epoch = self.quant.quantize_up(sector.expiration);
+            sectors_group.entry(epoch).or_insert(BitField::new()).set(sector.sector_number);
+            sectors_by_number.insert(sector.sector_number, sector);
+        }
+        for (epoch, sectors_field) in sectors_group {
+            let mut expiration_set = self.may_get(epoch)?;
+            let on_time_sectors = expiration_set.on_time_sectors.bitand(&sectors_field);
+            for sector_number in on_time_sectors.iter() {
+                let power =
+                    power_for_sector(sector_size, sectors_by_number.get(&sector_number).unwrap());
+                // If the sector expires on-time at this epoch, leave it here but change faulty power to active.
+                // The pledge is already part of the on-time pledge at this entry.
+                expiration_set.faulty_power -= &power;
+                expiration_set.active_power += &power;
+
+                recovered_power += &power;
+                remaining.remove(&sector_number);
+            }
+            let mut early_sectors = expiration_set.early_sectors.bitand(&sectors_field);
```

Comment:
We can probably optimize this by checking if `expiration_set.early_sectors` is empty up-front. It usually will be.